### ISSUE TITLE
Bug Fix: Ensure ServletRequestListener#requestDestroyed is always called

### ISSUE THREAD

**Issue by brideck** [OpenLiberty/open-liberty#19227]
**Describe the bug**  
The ServletRequestListener#requestDestroyed call does not occur.

**Steps to Reproduce**  
The two conditions must be met: the webContainer property deferServletRequestListenerDestroyOnError is true and an appSecurity-x.0 feature is enabled. 

**Expected behavior**  
ServletRequestListener#requestDestroyed method should be called when a request goes out of scope. 

**Diagnostic information:**  
 - OpenLiberty Version: All of Liberty
 - Affected feature(s) servlet-4.0, appSecurity-4.0
 - Java Version: 
 - server.xml configuration: `<webContainer deferServletRequestListenerDestroyOnError="true"/>`

**Additional context**  
This issue is related to #18281
Original Issue Text: 

With the inclusion of PR https://github.com/OpenLiberty/open-liberty/pull/17210, the org.jboss.cdi.tck.tests.context.conversation.determination.ConversationDeterminationTest has started to fail intermittently in our EE9 CTS testing with the following exception:

```
org.jboss.weld.contexts.BusyConversationException: WELD-000322: Conversation lock timed out: foo
    at org.jboss.weld.contexts.AbstractConversationContext.initialize(AbstractConversationContext.java:239)
    at org.jboss.weld.module.web.context.http.LazyHttpConversationContextImpl.initialize(LazyHttpConversationContextImpl.java:86)
    at org.jboss.weld.module.web.context.http.LazyHttpConversationContextImpl.checkContextInitialized(LazyHttpConversationContextImpl.java:121)
    at org.jboss.weld.contexts.AbstractConversationContext.getCurrentConversation(AbstractConversationContext.java:508)
    at org.jboss.weld.bean.builtin.ConversationBean.newInstance(ConversationBean.java:31)
    at org.jboss.weld.bean.builtin.ConversationBean.newInstance(ConversationBean.java:16)
    at org.jboss.weld.bean.builtin.AbstractDecorableBuiltInBean.create(AbstractDecorableBuiltInBean.java:53)
    at org.jboss.weld.contexts.AbstractContext.get(AbstractContext.java:96)
    at org.jboss.weld.bean.ContextualInstanceStrategy$DefaultContextualInstanceStrategy.get(ContextualInstanceStrategy.java:100)
    at org.jboss.weld.bean.ContextualInstance.get(ContextualInstance.java:50)
    at org.jboss.weld.bean.proxy.ContextBeanInstance.getInstance(ContextBeanInstance.java:102)
    at org.jboss.weld.bean.proxy.ProxyMethodHandler.invoke(ProxyMethodHandler.java:105)
    at jakarta.enterprise.context.Conversation$Proxy$_$$_WeldClientProxy.getId(Unknown Source)
    at org.jboss.cdi.tck.tests.context.conversation.determination.QuxAsyncListener.checkSameConversationActive(QuxAsyncListener.java:65)
    at org.jboss.cdi.tck.tests.context.conversation.determination.QuxAsyncListener.onStartAsync(QuxAsyncListener.java:61)
    at com.ibm.ws.webcontainer.async.AsyncListenerEntry.invokeOnStartAsync(AsyncListenerEntry.java:112)
    at com.ibm.ws.webcontainer.async.AsyncContextImpl.initialize(AsyncContextImpl.java:818)
    at com.ibm.ws.webcontainer.srt.SRTServletRequest.startAsync(SRTServletRequest.java:3663)
    at com.ibm.ws.webcontainer.srt.SRTServletRequest.startAsync(SRTServletRequest.java:3619)
    at org.jboss.cdi.tck.tests.context.conversation.determination.AsyncFooServlet.service(AsyncFooServlet.java:60)
    at jakarta.servlet.http.HttpServlet.service(HttpServlet.java:587)
    at com.ibm.ws.webcontainer.servlet.ServletWrapper.service(ServletWrapper.java:1258)
    at com.ibm.ws.webcontainer.servlet.ServletWrapper.handleRequest(ServletWrapper.java:746)
    at com.ibm.ws.webcontainer.servlet.ServletWrapper.handleRequest(ServletWrapper.java:443)
    at com.ibm.ws.webcontainer.filter.WebAppFilterChain.invokeTarget(WebAppFilterChain.java:193)
    at com.ibm.ws.webcontainer.filter.WebAppFilterChain.doFilter(WebAppFilterChain.java:98)
    at com.ibm.ws.security.jaspi.JaspiServletFilter.doFilter(JaspiServletFilter.java:56)
    at com.ibm.ws.webcontainer.filter.FilterInstanceWrapper.doFilter(FilterInstanceWrapper.java:201)
    at com.ibm.ws.webcontainer.filter.WebAppFilterChain.doFilter(WebAppFilterChain.java:91)
    at com.ibm.ws.webcontainer.filter.WebAppFilterManager.doFilter(WebAppFilterManager.java:1002)
    at com.ibm.ws.webcontainer.filter.WebAppFilterManager.invokeFilters(WebAppFilterManager.java:1140)
    at com.ibm.ws.webcontainer.filter.WebAppFilterManager.invokeFilters(WebAppFilterManager.java:1011)
    at com.ibm.ws.webcontainer.webapp.WebAppRequestDispatcher.dispatch(WebAppRequestDispatcher.java:1411)
    at com.ibm.ws.webcontainer.async.DispatchRunnable.run(DispatchRunnable.java:90)
    at com.ibm.ws.webcontainer.async.ServiceWrapper.wrapAndRun(ServiceWrapper.java:236)
    at com.ibm.ws.webcontainer.async.ContextWrapper.run(ContextWrapper.java:28)
    at com.ibm.ws.threading.internal.ExecutorServiceImpl$RunnableWrapper.run(ExecutorServiceImpl.java:238)
    at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
    at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
    at java.base/java.lang.Thread.run(Thread.java:866)
```

This was seen previously as a hard failure in pre-integration testing where it was partially fixed by PR https://github.com/OpenLiberty/open-liberty/pull/18333, but there is still more work to be done.

**Comment by volosied** [comment]
I'm swamped with other work, and I haven't been able to complete my investigation.  I will come back to this next Monday.

**Comment by volosied** [comment]
Looked over this today. I think I know why it's failing?  At least, I've been able to reproduce this error locally when I enabled the appSecurity-4.0 feature. However, the issue I found here affects both the passing and the failing cases.  I think it's just a coincidence that it passes in the "good case" -- this stumped me for a while. 

During the test, an asynchronous servlet throws an exception and while Liberty tries to handle it, it messes up by not calling ServletRequestListener.requestDestroyed(). 

Basically, it related to the `isCDINeeded` check below.  If true, then the servletRequestListenersNotificationNeeded is set to false, and that requestDestroyed calls are not invoked (and the CDI conversation scope is not unlocked on the next request). They are not invoked because "collabMetaData.setServletRequestCreated(true);" needs to occur for those ServletListeners  destroy methods to be called later on.  This is a bit similar to 18333, but in a slightly different scenario.


Both cases evaluate `isCDINeeded` to true. Not sure why it doesn't fail in both scenarios, however...? 

Bad Case:
```
[10/27/21, 10:01:46:176 PDT] 00000027 id=1f2af1ac .ibm.ws.webcontainer.security.WebAppSecurityCollaboratorImpl > isCDINeeded Entry 
[10/27/21, 10:01:46:176 PDT] 00000027 id=00000000 com.ibm.ws.webcontainer.osgi.WebContainer                    3 getServletContainerSpecLevel 
                                                                                                               loadedContainerSpecLevel [ 50 ]
[10/27/21, 10:01:46:176 PDT] 00000027 id=1f2af1ac .ibm.ws.webcontainer.security.WebAppSecurityCollaboratorImpl < isCDINeeded Exit  
                                                                                                               true
```

"Good Trace" (Again, not exactly sure why it passes. Somehow the same thread containing the lock avoids the BusyConversationException error ...?): 
```
[10/27/21, 9:59:13:468 PDT] 00000029 id=1f511698 .ibm.ws.webcontainer.security.WebAppSecurityCollaboratorImpl > isCDINeeded Entry 
[10/27/21, 9:59:13:468 PDT] 00000029 id=00000000 com.ibm.ws.webcontainer.osgi.WebContainer                    3 getServletContainerSpecLevel 
                                                                                                               loadedContainerSpecLevel [ 50 ]
[10/27/21, 9:59:13:468 PDT] 00000029 id=1f511698 .ibm.ws.webcontainer.security.WebAppSecurityCollaboratorImpl < isCDINeeded Exit  
                                                                                                               true
```


Here is the isCDINeeded check and where **servletRequestListenersNotificationNeeded** is set to false: 
https://github.com/OpenLiberty/open-liberty/blob/f1bcfd58b46f4e3d5b306159c0b544dca3e7b3fe/dev/com.ibm.ws.webcontainer/src/com/ibm/wsspi/webcontainer/collaborator/CollaboratorHelper.java#L453-L457

A few lines down, collabMetaData.setServletRequestCreated(true); should be called in order to have the ServletRequestListeners invoked later on during the `postInvokeCollaborators` method call. 
https://github.com/OpenLiberty/open-liberty/blob/f1bcfd58b46f4e3d5b306159c0b544dca3e7b3fe/dev/com.ibm.ws.webcontainer/src/com/ibm/wsspi/webcontainer/collaborator/CollaboratorHelper.java#L491-L503

CDI Check in Question.  I've been able to reproduce this error locally when I enabled the `appSecurity-4.0` features. 
https://github.com/OpenLiberty/open-liberty/blob/1861462bd27ff3c07067f3c543fe9487276dffe4/dev/com.ibm.ws.webcontainer.security/src/com/ibm/ws/webcontainer/security/WebAppSecurityCollaboratorImpl.java#L477-L489
